src/Concretize.hs

```diff
@@ -436,43 +439,41 @@ concretizeType _ ft@FuncTy {} =
   if isTypeGeneric ft
     then Right []
     else Right [defineFunctionTypeAlias ft]
-concretizeType typeEnv arrayTy@(StructTy (ConcreteNameTy "Array") varTys) =
+concretizeType typeEnv arrayTy@(StructTy (ConcreteNameTy (SymPath [] "Array")) varTys) =
   if isTypeGeneric arrayTy
     then Right []
     else do
       deps <- mapM (concretizeType typeEnv) varTys
       Right (defineArrayTypeAlias arrayTy : concat deps)
 -- TODO: Remove ugly duplication of code here:
-concretizeType typeEnv arrayTy@(StructTy (ConcreteNameTy "StaticArray") varTys) =
+concretizeType typeEnv arrayTy@(StructTy (ConcreteNameTy (SymPath [] "StaticArray")) varTys) =
   if isTypeGeneric arrayTy
     then Right []
     else do
       deps <- mapM (concretizeType typeEnv) varTys
       Right (defineStaticArrayTypeAlias arrayTy : concat deps)
--- TODO: handle polymorphic constructors (a b)
-concretizeType typeEnv genericStructTy@(StructTy (ConcreteNameTy name) _) =
-  case lookupInEnv (SymPath lookupPath structName) (getTypeEnv typeEnv) of
-    Just (_, Binder _ (XObj (Lst (XObj (Deftype originalStructTy) _ _ : _ : rest)) _ _)) ->
+concretizeType typeEnv genericStructTy@(StructTy (ConcreteNameTy (SymPath _ name)) _) =
+  -- TODO: This function only looks up direct children of the type environment.
+  -- However, spath can point to types that belong to a module. Pass the global env here.
+  case (getTypeBinder typeEnv name) of
+    Right (Binder _ x) -> go x
```

Comment:
What does `go` stand for?

Reply:
Nothing in particular. It's just a haskell convention for local function definitions: https://stackoverflow.com/questions/5844653/haskell-why-the-convention-to-name-a-helper-function-go. Admittedly it's also a function of my relative laziness while writing.